substrate/frame/sassafras/src/lib.rs

```diff
@@ -0,0 +1,1059 @@
+};
+use frame_system::{
+	offchain::{SendTransactionTypes, SubmitTransaction},
+	pallet_prelude::BlockNumberFor,
+};
+use sp_consensus_sassafras::{
+	digests::{ConsensusLog, NextEpochDescriptor, SlotClaim},
+	vrf, AuthorityId, Epoch, EpochConfiguration, Randomness, Slot, TicketBody, TicketEnvelope,
+	TicketId, RANDOMNESS_LENGTH, SASSAFRAS_ENGINE_ID,
+};
+use sp_io::hashing;
+use sp_runtime::{generic::DigestItem, traits::One, BoundToRuntimeAppPublic};
+use sp_std::prelude::Vec;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+#[cfg(all(feature = "std", test))]
+mod mock;
+#[cfg(all(feature = "std", test))]
+mod tests;
+
+pub mod weights;
+pub use weights::WeightInfo;
+
+pub use pallet::*;
+
+const LOG_TARGET: &str = "sassafras::runtime 🌳";
+
+// Contextual string used by the VRF to generate per-block randomness.
+const RANDOMNESS_VRF_CONTEXT: &[u8] = b"SassafrasRandomness";
+
+// Max length for segments holding unsorted tickets.
+const SEGMENT_MAX_SIZE: u32 = 128;
+
+// Convenience type
+type AuthoritiesVec<T> = WeakBoundedVec<AuthorityId, <T as Config>::MaxAuthorities>;
+
+/// Tickets metadata.
+#[derive(Debug, Default, PartialEq, Encode, Decode, TypeInfo, MaxEncodedLen, Clone, Copy)]
+pub struct TicketsMetadata {
+	/// Number of outstanding next epoch tickets requiring to be sorted.
+	///
+	/// These tickets are held by the [`UnsortedSegments`] storage map in segments
+	/// containing at most `SEGMENT_MAX_SIZE` items.
+	pub unsorted_tickets_count: u32,
+
+	/// Number of tickets available for current and next epoch.
+	///
+	/// These tickets are held by the [`TicketsIds`] storage map.
+	///
+	/// The array entry to be used for the current epoch is computed as epoch index modulo 2.
+	pub tickets_count: [u32; 2],
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	/// The Sassafras pallet.
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	/// Configuration parameters.
+	#[pallet::config]
+	pub trait Config: frame_system::Config + SendTransactionTypes<Call<Self>> {
+		/// Amount of slots that each epoch should last.
+		#[pallet::constant]
+		type EpochLength: Get<u64>;
+
+		/// Max number of authorities allowed.
+		#[pallet::constant]
+		type MaxAuthorities: Get<u32>;
+
+		/// Epoch change trigger.
+		///
+		/// Logic to be triggered on every block to query for whether an epoch has ended
+		/// and to perform the transition to the next epoch.
+		type EpochChangeTrigger: EpochChangeTrigger;
+
+		/// Weight information for all calls of this pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	/// Max number of tickets allowed by the configuration.
+	///
+	/// In practice trims down the `Config::EpochLength` value to at most u32::MAX.
+	pub struct MaxTicketsFor<T: Config>(sp_std::marker::PhantomData<T>);
+
+	impl<T: Config> Get<u32> for MaxTicketsFor<T> {
+		fn get() -> u32 {
+			T::EpochLength::get().try_into().unwrap_or(u32::MAX)
```

Comment:
Can't we make `EpochLength` a `u32` then?

Reply:
Well, I guess yes. IIRC I made it an `u64` when I was initially following the Babe's code.

WDYT? 4 billions of slots enough for an epoch? Probably in practice yes 😃

Why Babe used a `u64` for epoch length? 

I don't know, there is still this damn voice in my head that asks me to keep it equal to Babe 🤣

Reply:
> Why Babe used a `u64` for epoch length?

Probably because I wasn't using my brain 😅 You shouldn't make the same mistake. I think the bigger point here is that both things should have the same type. Using `u64` for `MaxTicketsFor` also works (would that be practical though?).

Reply:
I use `MaxTicketsFor<T>` mostly as a bound for the `BoundedVec`s. This bound is required to implement `Get<u32>` (this is why I implemented this saturated `Get<u32>` for a u64).

Let's say that this allows to have a longer epoch length but that is impossible to completelly fill with tickets because you are bounded to 4 billions.

Maybe this doesn't make any sense. I change the epoch length to `u32` :-/

Reply:
Done 41a34ff038148754c71f89a46aaa841298992621